File: Samples/DeviceEnumerationAndPairing/js/js/scenario9_custompairdevice.js
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";

    var DevEnum = Windows.Devices.Enumeration;
    var resultCollection = new WinJS.Binding.List([]);
    var resultsListView;
    var deviceWatcherHelper = new SdkSample.DeviceWatcherHelper(resultCollection);
    var providePinTaskSrc;
    var confirmPinTaskSrc;
    var providePasswordCredentialTaskSrc;

    var page = WinJS.UI.Pages.define("/html/scenario9_custompairdevice.html", {
        ready: function (element, options) {

            // Hook up button event handlers
            document.getElementById("startWatcherButton").addEventListener("click", startWatcher, false);
            document.getElementById("stopWatcherButton").addEventListener("click", stopWatcher, false);
            document.getElementById("pairButton").addEventListener("click", pairDevice, false);
            document.getElementById("unpairButton").addEventListener("click", unpairDevice, false);
            document.getElementById("okButton").addEventListener("click", okButton, false);
            document.getElementById("yesButton").addEventListener("click", yesButton, false);
            document.getElementById("noButton").addEventListener("click", noButton, false);
            document.getElementById("verifyButton").addEventListener("click", verifyButton, false);

            // Hook up result list selection changed event handler
            resultsListView = element.querySelector("#resultsListView").winControl;
            resultsListView.addEventListener("selectionchanged", onSelectionChanged);
            // Hook up result list data binding
            resultsListView.itemDataSource = resultCollection.dataSource;

            // We need to do some extra work when an item changes.
            // This event is raised by the DeviceWatcherHelper.
            resultCollection.addEventListener("deviceupdated", onDeviceUpdated);

            // Manually bind selector options
            DisplayHelpers.pairingSelectors.forEach(function each(item) {
                var option = document.createElement("option");
                option.textContent = item.displayName;
                selectorComboBox.appendChild(option);
            });
            selectorComboBox.selectedIndex = 0;

            // Process any data bindings
            WinJS.UI.processAll();
        },
        unload: function () {
            deviceWatcherHelper.reset();
            resultCollection.splice(0, resultCollection.length);
        }
    });

    function startWatcher() {
        startWatcherButton.disabled = true;
        resultCollection.splice(0, resultCollection.length);

        // Get the device selector chosen by the UI then add additional constraints for devices that 
        // can be paired or are already paired.
        var selectedItem = DisplayHelpers.pairingSelectors.getAt(selectorComboBox.selectedIndex);
        var selector = "(" + selectedItem.selector + ")" + " AND (System.Devices.Aep.CanPair:=System.StructuredQueryType.Boolean#True OR System.Devices.Aep.IsPaired:=System.StructuredQueryType.Boolean#True)";

        var deviceWatcher;
        if (selectedItem.kind == DevEnum.DeviceInformationKind.unknown) {
            // Kind will be determined by the selector
            deviceWatcher = DevEnum.DeviceInformation.createWatcher(
                selector,
                null // don't request additional properties for this sample
                );
        }
        else {
            // Kind will be determined by the selector
            deviceWatcher = DevEnum.DeviceInformation.createWatcher(
                selector,
                null, // don't request additional properties for this sample
                selectedItem.kind);
        }

        WinJS.log && WinJS.log("Starting watcher...", "sample", "status");
        deviceWatcherHelper.startWatcher(deviceWatcher);
        stopWatcherButton.disabled = false;
    }

    function stopWatcher() {
        stopWatcherButton.disabled = true;
        deviceWatcherHelper.stopWatcher();
        startWatcherButton.disabled = false;
    }

    function getSelectedItem() {
        var selectedIndices = resultsListView.selection.getIndices();
        if (selectedIndices.length > 0) {
            return resultCollection.getAt(selectedIndices[0]);
        }
    }

    function onDeviceUpdated(e) {
        var id = e.detail.id;
        // If the item being updated is currently "selected", then update the pairing buttons
        var selectedItem = getSelectedItem();
        if (selectedItem && selectedItem.id === id) {
            UpdatePairingButtons();
        }
    }

    function pairDevice() {
        // Gray out the pair button and results view while pairing is in progress.
        pairButton.disabled = true;
        WinJS.log && WinJS.log("Pairing started. Please wait...", "sample", "status");

        var deviceDispInfo = getSelectedItem();
        if (deviceDispInfo) {
            var ceremoniesSelected = GetSelectedCeremonies();
            var protectionLevel = protectionLevelComboBox.selectedIndex;

            var customPairing = deviceDispInfo.deviceInfo.pairing.custom;
            customPairing.addEventListener("pairingrequested", pairingRequestHandler);
            
            customPairing.pairAsync(ceremoniesSelected, protectionLevel).done(
                function (pairingResult) {
                    var statusType = (pairingResult.status == DevEnum.DevicePairingResultStatus.paired) ? "status" : "error";
                    var message = (pairingResult.status == DevEnum.DevicePairingResultStatus.paired) ? "Paired" : "NotPaired";
                    WinJS.log && WinJS.log("Pairing result = " + message, "sample", statusType);

                    customPairing.removeEventListener("pairingrequested", pairingRequestHandler);

                    hidePairingPanel();
                    updatePairingButtons();
                });
        }
    }

    function unpairDevice() {
        // Gray out the pair button and results view while pairing is in progress.
        unpairButton.disabled = true;
        WinJS.log && WinJS.log("Unpairing started. Please wait...", "sample", "status");

        var deviceDispInfo = getSelectedItem();
        if (deviceDispInfo) {
            unpairButton.disabled = true;
            deviceDispInfo.deviceInfo.pairing.unpairAsync().done(
                function (unpairingResult) {
                    var message = (unpairingResult.status == DevEnum.DeviceUnpairingResultStatus.unpaired) ? "Unpaired" : "Failed";
                    var statusType = (unpairingResult.status == DevEnum.DeviceUnpairingResultStatus.unpaired) ? "status" : "error";
                    WinJS.log && WinJS.log("Unpairing result = " + message, "sample", statusType);

                    updatePairingButtons();
                });
        }
    }

    function pairingRequestHandler(args) {
        if (args.pairingKind == DevEnum.DevicePairingKinds.confirmOnly) {
            // Windows itself will pop the confirmation dialog as part of "consent" if this is running on Desktop or Mobile
            // If this is an App for 'Windows IoT Core' where there is no Windows Consent UX, you may want to provide your own confirmation.
            args.accept();
        }
        else if (args.pairingKind == DevEnum.DevicePairingKinds.displayPin) {
            // We just show the PIN on this side. The ceremony is actually completed when the user enters the PIN
            // on the target device. We automatically accept here since we can't really "cancel" the operation
            // from this side.
            args.accept();

            showPairingPanel("Please enter this PIN on the device you are pairing with: " + args.pin, args.pairingKind);
        }
        else if (args.pairingKind == DevEnum.DevicePairingKinds.providePin) {
            // A PIN may be shown on the target device and the user needs to enter the matching PIN on 
            // this Windows device. Get a deferral so we can perform the async request to the user.
            var collectPinDeferral = args.getDeferral();

            getPinFromUserAsync().then(function (pin) {
                if (pin) {
                    args.accept(pin);
                }
                collectPinDeferral.complete();
            });
        }
        else if (args.pairingKind == DevEnum.DevicePairingKinds.confirmPinMatch) {
            // We show the PIN here and the user responds with whether the PIN matches what they see
            // on the target device. Response comes back and we set it on the PinComparePairingRequestedData
            // then complete the deferral.
            var displayMessageDeferral = args.getDeferral();

            getUserConfirmationAsync(args.pin).then(function (accept) {
                if (accept) {
                    args.accept();
                }
                displayMessageDeferral.complete();
            });
        }
        else if (args.pairingKind == DevEnum.DevicePairingKinds.providePasswordCredential) {
            var collectCredentialDeferral = args.getDeferral();
            getPasswordCredentialFromUserAsync().then(function (credential) {
                if (credential) {
                    args.acceptWithPasswordCredential(credential);
                }
                collectCredentialDeferral.complete();
            });
        }
    }

    function showPairingPanel(text, pairingKind) {
        document.getElementById("pairingPanel").style.display = "block";
        document.getElementById("pinEntryTextBox").style.display = "none";
        document.getElementById("okButton").style.display = "none";
        document.getElementById("yesButton").style.display = "none";
        document.getElementById("noButton").style.display = "none";
        document.getElementById("usernameEntryTextBox").style.display = "none";
        document.getElementById("passwordEntryTextBox").style.display = "none";
        document.getElementById("verifyButton").style.display = "none";
        pairingTextBlock.innerHTML = text;

        switch (pairingKind) {
            case DevEnum.DevicePairingKinds.confirmOnly:
            case DevEnum.DevicePairingKinds.displayPin:
                // Don't need any buttons
                break;
            case DevEnum.DevicePairingKinds.providePin:
                pinEntryTextBox.value = "";
                document.getElementById("pinEntryTextBox").style.display = "inline";
                document.getElementById("okButton").style.display = "inline";
                break;
            case DevEnum.DevicePairingKinds.confirmPinMatch:
                document.getElementById("yesButton").style.display = "inline";
                document.getElementById("noButton").style.display = "inline";
                break;
            case DevEnum.DevicePairingKinds.providePasswordCredential:
                document.getElementById("usernameEntryTextBox").style.display = "inline";
                document.getElementById("passwordEntryTextBox").style.display = "inline";
                document.getElementById("verifyButton").style.display = "inline";
                break;
        }
    }

    function hidePairingPanel() {
        document.getElementById("pairingPanel").style.display = "none";
        pairingTextBlock.innerHTML = "";
    }

    function getPinFromUserAsync() {
        hidePairingPanel();
        completeProvidePinTask();

        showPairingPanel(
            "Please enter the PIN shown on the device you're pairing with",
            DevEnum.DevicePairingKinds.providePin);

        return new WinJS.Promise(function (c) {
            providePinTaskSrc = c;
        });
    }

    // If pin is not provided, then any pending pairing request is abandoned.
    function completeProvidePinTask(pin) {
        if (providePinTaskSrc) {
            providePinTaskSrc(pin);
            providePinTaskSrc = null;
        }
    }

    function getPasswordCredentialFromUserAsync() {
        hidePairingPanel();
        completePasswordCredential(); // Abandon any previous pin request.

        showPairingPanel(
            "Please enter the username and password",
            DevEnum.DevicePairingKinds.providePasswordCredential);

        return new WinJS.Promise(function (c) {
            providePasswordCredentialTaskSrc = c;
        });
    }

    function  completePasswordCredential(username, password)
    {
        if (providePasswordCredentialTaskSrc) {
            if (username) {
                var passwordCredential = new Windows.Security.Credentials.PasswordCredential();
                passwordCredential.userName = username;
                passwordCredential.password = password;
                providePasswordCredentialTaskSrc(passwordCredential);
            } else {
                providePasswordCredentialTaskSrc(null);
            }
            providePasswordCredentialTaskSrc = null;
        }
    }

    function getUserConfirmationAsync(pin) {
        hidePairingPanel();
        completeConfirmPinTask(false);

        showPairingPanel(
            "Does the following PIN match the one shown on the device you are pairing?: " + pin,
            DevEnum.DevicePairingKinds.confirmPinMatch);

        return new WinJS.Promise(function (c) {
            confirmPinTaskSrc = c;
        });
    }

    // If pin is not provided, then any pending pairing request is abandoned.
    function completeConfirmPinTask(accept) {
        if (confirmPinTaskSrc) {
            confirmPinTaskSrc(accept);
            confirmPinTaskSrc = null;
        }
    }
    
    function okButton() {
        // OK button is only used for the ProvidePin scenario
        completeProvidePinTask(pinEntryTextBox.value);
        hidePairingPanel();
    }

    function verifyButton(){
        // verify button is only used for the ProvidePin scenario
        completePasswordCredential(usernameEntryTextBox.value, passwordEntryTextBox.value);
        hidePairingPanel();
    }

    function yesButton() {
        completeConfirmPinTask(true);
        hidePairingPanel();
    }

    function noButton() {
        completeConfirmPinTask(false);
        hidePairingPanel();
    }

    function GetSelectedCeremonies() {
        var ceremonySelection = DevEnum.DevicePairingKinds.none;

        WinJS.Utilities.query(".win-checkbox").forEach(function (checkbox) {
            if (checkbox.checked) {
                switch (checkbox.id) {
                    case "confirmOnlyOption":
                        ceremonySelection |= DevEnum.DevicePairingKinds.confirmOnly;
                        break;
                    case "displayPinOption":
                        ceremonySelection |= DevEnum.DevicePairingKinds.displayPin;
                        break;
                    case "providePinOption":
                        ceremonySelection |= DevEnum.DevicePairingKinds.providePin;
                        break;
                    case "confirmPinMatchOption":
                        ceremonySelection |= DevEnum.DevicePairingKinds.confirmPinMatch;
                        break;
                    case "passwordCredentialOption":
                        ceremonySelection |= DevEnum.DevicePairingKinds.providePasswordCredential;
                        break;
                }
            }
        });

        return ceremonySelection;
    }

    function onSelectionChanged() {
        updatePairingButtons();
    }

    function updatePairingButtons() {

        var deviceDispInfo = getSelectedItem();

        if (deviceDispInfo &&
            deviceDispInfo.canPair &&
            !deviceDispInfo.isPaired) {
            pairButton.disabled = false;
        }
        else {
            pairButton.disabled = true;
        }

        if (deviceDispInfo &&
            deviceDispInfo.isPaired) {
            unpairButton.disabled = false;
        }
        else {
            unpairButton.disabled = true;
        }
    }
})();

File: Samples/DeviceEnumerationAndPairing/cppwinrt/Scenario9_CustomPairDevice.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario9_CustomPairDevice.h"
#include "DeviceHelpers.h"
#include "SampleConfiguration.h"
#include "Scenario9_CustomPairDevice.g.cpp"

using namespace winrt;
using namespace winrt::Windows::Devices::Enumeration;
using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::Security::Credentials;
using namespace winrt::Windows::UI::Xaml;
using namespace winrt::Windows::UI::Xaml::Controls;
using namespace winrt::Windows::UI::Xaml::Navigation;

namespace winrt::SDKTemplate::implementation
{
    Scenario9_CustomPairDevice::Scenario9_CustomPairDevice()
    {
        InitializeComponent();
        deviceWatcherHelper->DeviceChanged({ get_weak(), &Scenario9_CustomPairDevice::OnDeviceListChanged });
    }

    void Scenario9_CustomPairDevice::OnNavigatedTo(NavigationEventArgs const&)
    {
        resultsListView().ItemsSource(resultCollection);

        selectorComboBox().ItemsSource(DeviceSelectorChoices::PairingSelectors());
        selectorComboBox().SelectedIndex(0);

        protectionLevelComboBox().ItemsSource(ProtectionSelectorChoices::Selectors());
        protectionLevelComboBox().SelectedIndex(0);
    }

    void Scenario9_CustomPairDevice::OnNavigatedFrom(NavigationEventArgs const&)
    {
        deviceWatcherHelper->Reset();
        CompleteProvidePinTask(); // Abandon any previous pin request.
        CompletePasswordCredential(); // Abandon any previous pin request.
        CompleteConfirmPinTask(false); // Abandon any previous request.
    }

    void Scenario9_CustomPairDevice::StartWatcherButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        StartWatcher();
    }

    void Scenario9_CustomPairDevice::StopWatcherButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        StopWatcher();
    }

    void Scenario9_CustomPairDevice::StartWatcher()
    {
        startWatcherButton().IsEnabled(false);
        resultCollection.Clear();

        // Get the device selector chosen by the UI then add additional constraints for devices that
        // can be paired or are already paired.
        DeviceSelectorInfo* deviceSelectorInfo = FromInspectable<DeviceSelectorInfo>(selectorComboBox().SelectedItem());
        hstring selector = L"(" + deviceSelectorInfo->Selector() + L") AND (System.Devices.Aep.CanPair:=System.StructuredQueryType.Boolean#True OR System.Devices.Aep.IsPaired:=System.StructuredQueryType.Boolean#True)";

        DeviceWatcher deviceWatcher{ nullptr };
        if (deviceSelectorInfo->Kind() == DeviceInformationKind::Unknown)
        {
            // Kind will be determined by the selector
            deviceWatcher = DeviceInformation::CreateWatcher(
                selector,
                nullptr // don't request additional properties for this sample
            );
        }
        else
        {
            // Kind is specified in the selector info
            deviceWatcher = DeviceInformation::CreateWatcher(
                selector,
                nullptr, // don't request additional properties for this sample
                deviceSelectorInfo->Kind());
        }

        rootPage.NotifyUser(L"Starting Watcher...", NotifyType::StatusMessage);
        deviceWatcherHelper->StartWatcher(deviceWatcher);
        stopWatcherButton().IsEnabled(true);
    }

    void Scenario9_CustomPairDevice::StopWatcher()
    {
        stopWatcherButton().IsEnabled(false);

        deviceWatcherHelper->StopWatcher();

        startWatcherButton().IsEnabled(true);
    }

    void Scenario9_CustomPairDevice::OnDeviceListChanged(DeviceWatcher const&, hstring const& id)
    {
        // If the item being updated is currently "selected", then update the pairing buttons
        auto item = resultsListView().SelectedItem();
        DeviceInformationDisplay* deviceInformationDisplay = FromInspectable<DeviceInformationDisplay>(item);
        if ((deviceInformationDisplay != nullptr) && (deviceInformationDisplay->Id() == id))
        {
            UpdatePairingButtons();
        }
    }

    fire_and_forget Scenario9_CustomPairDevice::PairButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        PasswordCredential credential = co_await GetPasswordCredentialFromUserAsync();
        if (credential)
        {
            rootPage.NotifyUser(L"Entered", NotifyType::StatusMessage);
        }

        // Gray out the pair button and results view while pairing is in progress.
        resultsListView().IsEnabled(false);
        pairButton().IsEnabled(false);
        rootPage.NotifyUser(L"Pairing started. Please wait...", NotifyType::StatusMessage);

        // Get the device selected for pairing
        IInspectable item = resultsListView().SelectedItem();
        DeviceInformationDisplay* deviceInfoDisp = FromInspectable<DeviceInformationDisplay>(item);

        // Get ceremony type and protection level selections
        DevicePairingKinds ceremoniesSelected = GetSelectedCeremonies();
        IInspectable protectionItem = protectionLevelComboBox().SelectedItem();
        ProtectionLevelSelectorInfo* protectionLevelInfo = FromInspectable<ProtectionLevelSelectorInfo>(protectionItem);
        DevicePairingProtectionLevel protectionLevel = protectionLevelInfo->ProtectionLevel();

        DeviceInformationCustomPairing customPairing = deviceInfoDisp->DeviceInformation().Pairing().Custom();

        event_token token = customPairing.PairingRequested({ get_weak(), &Scenario9_CustomPairDevice::PairingRequestedHandler });
        DevicePairingResult result = co_await customPairing.PairAsync(ceremoniesSelected, protectionLevel);
        customPairing.PairingRequested(token);

        rootPage.NotifyUser(
            L"Pairing result = " + to_hstring(result.Status()),
            result.Status() == DevicePairingResultStatus::Paired ? NotifyType::StatusMessage : NotifyType::ErrorMessage);

        HidePairingPanel();
        UpdatePairingButtons();
        resultsListView().IsEnabled(true);
    }

    fire_and_forget Scenario9_CustomPairDevice::UnpairButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        // Gray out the pair button and results view while unpairing is in progress.
        resultsListView().IsEnabled(false);
        unpairButton().IsEnabled(false);
        rootPage.NotifyUser(L"Unpairing started. Please wait...", NotifyType::StatusMessage);

        IInspectable item = resultsListView().SelectedItem();
        DeviceInformationDisplay* deviceInfoDisp = FromInspectable<DeviceInformationDisplay>(item);

        DeviceUnpairingResult result = co_await deviceInfoDisp->DeviceInformation().Pairing().UnpairAsync();

        rootPage.NotifyUser(
            L"Unpairing result = " + to_hstring(result.Status()),
            result.Status() == DeviceUnpairingResultStatus::Unpaired ? NotifyType::StatusMessage : NotifyType::ErrorMessage);

        UpdatePairingButtons();
        resultsListView().IsEnabled(true);
    }

    void Scenario9_CustomPairDevice::ResultsListView_SelectionChanged(IInspectable const&, SelectionChangedEventArgs const&)
    {
        UpdatePairingButtons();
    }

    fire_and_forget Scenario9_CustomPairDevice::PairingRequestedHandler(DeviceInformationCustomPairing const&, DevicePairingRequestedEventArgs args)
    {
        auto lifetime = get_strong();

        switch (args.PairingKind())
        {
        case DevicePairingKinds::ConfirmOnly:
            // Windows itself will pop the confirmation dialog as part of "consent" if this is running on Desktop or Mobile
            // If this is an App for 'Windows IoT Core' where there is no Windows Consent UX, you may want to provide your own confirmation.
            args.Accept();
            break;

        case DevicePairingKinds::DisplayPin:
            // We just show the PIN on this side. The ceremony is actually completed when the user enters the PIN
            // on the target device. We automatically accept here since we can't really "cancel" the operation
            // from this side.
            args.Accept();

            // No need for a deferral since we don't need any decision from the user
            co_await resume_foreground(Dispatcher());

            ShowPairingPanel(
                L"Please enter this PIN on the device you are pairing with: " + args.Pin(),
                args.PairingKind());

            break;

        case DevicePairingKinds::ProvidePin:
        {
            // A PIN may be shown on the target device and the user needs to enter the matching PIN on
            // this Windows device. Get a deferral so we can perform the async request to the user.
            auto collectPinDeferral = args.GetDeferral();

            co_await resume_foreground(Dispatcher());

            hstring pin = co_await GetPinFromUserAsync();
            if (!pin.empty())
            {
                args.Accept(pin);
            }

            collectPinDeferral.Complete();
            break;
        }

        case DevicePairingKinds::ProvidePasswordCredential:
        {
            auto collectCredentialDeferral = args.GetDeferral();
            co_await resume_foreground(Dispatcher());

            PasswordCredential credential = co_await GetPasswordCredentialFromUserAsync();
            if (credential != nullptr)
            {
                args.AcceptWithPasswordCredential(credential);
            }
            collectCredentialDeferral.Complete();
            break;
        }

        case DevicePairingKinds::ConfirmPinMatch:
        {
            // We show the PIN here and the user responds with whether the PIN matches what they see
            // on the target device. Response comes back and we set it on the PinComparePairingRequestedData
            // then complete the deferral.
            auto displayMessageDeferral = args.GetDeferral();

            co_await resume_foreground(Dispatcher());
            bool accept = co_await GetUserConfirmationAsync(args.Pin());
            if (accept)
            {
                args.Accept();
            }

            displayMessageDeferral.Complete();
            break;
        }
        }
    }

    void Scenario9_CustomPairDevice::ShowPairingPanel(hstring const& text, DevicePairingKinds pairingKind)
    {
        pairingPanel().Visibility(Visibility::Collapsed);
        pinEntryTextBox().Visibility(Visibility::Collapsed);
        okButton().Visibility(Visibility::Collapsed);
        usernameEntryTextBox().Visibility(Visibility::Collapsed);
        passwordEntryTextBox().Visibility(Visibility::Collapsed);
        verifyButton().Visibility(Visibility::Collapsed);
        yesButton().Visibility(Visibility::Collapsed);
        noButton().Visibility(Visibility::Collapsed);
        pairingTextBlock().Text(text);

        switch (pairingKind)
        {
        case DevicePairingKinds::ConfirmOnly:
        case DevicePairingKinds::DisplayPin:
            // Don't need any buttons
            break;
        case DevicePairingKinds::ProvidePin:
            pinEntryTextBox().Text({});
            pinEntryTextBox().Visibility(Visibility::Visible);
            okButton().Visibility(Visibility::Visible);
            break;
        case DevicePairingKinds::ConfirmPinMatch:
            yesButton().Visibility(Visibility::Visible);
            noButton().Visibility(Visibility::Visible);
            break;
        case DevicePairingKinds::ProvidePasswordCredential:
            usernameEntryTextBox().Text({});
            passwordEntryTextBox().Text({});
            passwordEntryTextBox().Visibility(Visibility::Visible);
            usernameEntryTextBox().Visibility(Visibility::Visible);
            verifyButton().Visibility(Visibility::Visible);
            break;
        }

        pairingPanel().Visibility(Visibility::Visible);
    }

    void Scenario9_CustomPairDevice::HidePairingPanel()
    {
        pairingPanel().Visibility(Visibility::Collapsed);
        pairingTextBlock().Text({});
    }

    IAsyncOperation<hstring> Scenario9_CustomPairDevice::GetPinFromUserAsync()
    {
        HidePairingPanel();
        CompleteProvidePinTask(); // Abandon any previous pin request.

        ShowPairingPanel(
            L"Please enter the PIN shown on the device you're pairing with",
            DevicePairingKinds::ProvidePin);

        providePinTaskSrc = SimpleTaskCompletionSource<hstring>();

        return co_await providePinTaskSrc.Task();
    }

    // If pin is not provided, then any pending pairing request is abandoned.
    void Scenario9_CustomPairDevice::CompleteProvidePinTask(hstring pin)
    {
        providePinTaskSrc.SetResult(pin);
    }

    IAsyncOperation<PasswordCredential> Scenario9_CustomPairDevice::GetPasswordCredentialFromUserAsync()
    {
        HidePairingPanel();
        CompletePasswordCredential(); // Abandon any previous pin request.

        ShowPairingPanel(
            L"Please enter the username and password",
            DevicePairingKinds::ProvidePasswordCredential);

        providePasswordCredentialTaskSrc = SimpleTaskCompletionSource<PasswordCredential>();

        return co_await providePasswordCredentialTaskSrc.Task();
    }

    void Scenario9_CustomPairDevice::CompletePasswordCredential(hstring username, hstring password)
    {
        if (username.empty())// || password.empty())
        {
            providePasswordCredentialTaskSrc.SetResult(nullptr);
        }
        else
        {
            PasswordCredential passwordCredential;
            passwordCredential.UserName(username);
            passwordCredential.Password(password);
            providePasswordCredentialTaskSrc.SetResult(passwordCredential);
        }
    }

    IAsyncOperation<bool> Scenario9_CustomPairDevice::GetUserConfirmationAsync(hstring const& pin)
    {
        HidePairingPanel();
        CompleteConfirmPinTask(false); // Abandon any previous request.

        ShowPairingPanel(
            L"Does the following PIN match the one shown on the device you are pairing?: " + pin,
            DevicePairingKinds::ConfirmPinMatch);

        confirmPinTaskSrc = SimpleTaskCompletionSource<bool>();

        return co_await confirmPinTaskSrc.Task();
    }

    // If pin is not provided, then any pending pairing request is abandoned.
    void Scenario9_CustomPairDevice::CompleteConfirmPinTask(bool accept)
    {
        confirmPinTaskSrc.SetResult(accept);
    }


    void Scenario9_CustomPairDevice::okButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        // OK button is only used for the ProvidePin scenario
        CompleteProvidePinTask(pinEntryTextBox().Text());
        HidePairingPanel();
    }

    void Scenario9_CustomPairDevice::verifyButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        // verify button is only used for the ProvidePin scenario
        CompletePasswordCredential(usernameEntryTextBox().Text(), passwordEntryTextBox().Text());
        HidePairingPanel();
    }

    void Scenario9_CustomPairDevice::yesButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        CompleteConfirmPinTask(true);
        HidePairingPanel();
    }

    void Scenario9_CustomPairDevice::noButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        CompleteConfirmPinTask(false);
        HidePairingPanel();
    }

    DevicePairingKinds Scenario9_CustomPairDevice::GetSelectedCeremonies()
    {
        DevicePairingKinds ceremonySelection = DevicePairingKinds::None;
        if (confirmOnlyOption().IsChecked().Value()) ceremonySelection |= DevicePairingKinds::ConfirmOnly;
        if (displayPinOption().IsChecked().Value()) ceremonySelection |= DevicePairingKinds::DisplayPin;
        if (providePinOption().IsChecked().Value()) ceremonySelection |= DevicePairingKinds::ProvidePin;
        if (confirmPinMatchOption().IsChecked().Value()) ceremonySelection |= DevicePairingKinds::ConfirmPinMatch;
        if (passwordCredentialOption().IsChecked().Value()) ceremonySelection |= DevicePairingKinds::ProvidePasswordCredential;
        return ceremonySelection;
    }

    void Scenario9_CustomPairDevice::UpdatePairingButtons()
    {
        auto item = resultsListView().SelectedItem();
        DeviceInformationDisplay* deviceInfoDisp = FromInspectable<DeviceInformationDisplay>(item);
        if (deviceInfoDisp != nullptr &&
            deviceInfoDisp->DeviceInformation().Pairing().CanPair() &&
            !deviceInfoDisp->DeviceInformation().Pairing().IsPaired())
        {
            pairButton().IsEnabled(true);
        }
        else
        {
            pairButton().IsEnabled(false);
        }

        if (deviceInfoDisp != nullptr &&
            deviceInfoDisp->DeviceInformation().Pairing().IsPaired())
        {
            unpairButton().IsEnabled(true);
        }
        else
        {
            unpairButton().IsEnabled(false);
        }
    }
}

